lib/Sema/CSGen.cpp

```diff
@@ -3298,7 +3337,7 @@ namespace {
       CS.addConstraint(ConstraintKind::OptionalObject,
                        CS.getType(expr->getSubExpr()), objectTy,
                        locator);
-      return objectTy;
+      return addUnresolvedMemberChainConstraints(expr, objectTy, locator);
```

Comment:
We might be able, here and elsewhere, to lift this into the `ConstraintWalker` itself in `walkToExprPost` so that the logic for which expressions to "look through" for chains is contained entirely in `getMemberChainSubExpr`. That would require us to pay the cost of the `getMemberChainSubExpr(expr) == nullptr` on *every* expression, though, as opposed to just the ones that we know might be a part of an unresolved member chain. I think it would also require a new path element kind (or another function `getMemberResultLocator` that gives us the locator we should use to match the member result to the whole chain result)...